New display ST7735

SenexCrenshaw opened the pull request:
## Description:
Added oled display ST7735

**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#608

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

SenexCrenshaw committed:
Initial Commit - ST7735

SenexCrenshaw committed:
Initial Commit - ST7735

SenexCrenshaw committed:
Initial Commit - ST7735

SenexCrenshaw committed:
Initial Commit - ST7735

SenexCrenshaw committed:
Updated for CI checks

SenexCrenshaw committed:
Updated for CI checks

SenexCrenshaw committed:
Updated for CI checks

SenexCrenshaw committed:
Updated for CI checks

SenexCrenshaw committed:
Updated for travis build

SenexCrenshaw committed:
Updated for travis build

SenexCrenshaw committed:
Updated for travis build

SenexCrenshaw committed:
Updated for travis build

SenexCrenshaw committed:
Travis fixes

SenexCrenshaw committed:
Travis fixes

SenexCrenshaw committed:
Travis fixes

SenexCrenshaw committed:
Travis fixes

SenexCrenshaw committed:
Travis line too long

SenexCrenshaw committed:
Travis line too long

SenexCrenshaw committed:
Travis line too long

SenexCrenshaw committed:
Travis line too long

SenexCrenshaw committed:
Travis fixes

SenexCrenshaw committed:
Travis fixes

SenexCrenshaw committed:
Travis fixes

SenexCrenshaw committed:
Travis fixes

SenexCrenshaw committed:
Fixed up travis format issues

SenexCrenshaw committed:
Fixed up travis format issues

SenexCrenshaw committed:
Fixed up travis format issues

SenexCrenshaw committed:
Fixed up travis format issues

SenexCrenshaw committed:
Travis Fixes

SenexCrenshaw committed:
Travis Fixes

SenexCrenshaw committed:
Travis Fixes

SenexCrenshaw committed:
Travis Fixes

grinco commented:
Why isn't this being merged?

roberttco commented:
Is there any reason that the 7735 and 7789 support is not being added vies a vie the Addafruit libraries instead of rewriting it here and on other PRs ?  I think it makes sense to have the support be generic to those drivers rather than a specific flavor (e.g. #1050)  Wouldn't it me more useful to everyone ?

Pocke01 commented:
I Think maybe part of the answer is it this comment. 
https://github.com/esphome/feature-requests/issues/1#issuecomment-461822067

kbx81 commented:
@grinco PRs are often merged slowly, allowing them time to be tested and to mature. For example, we've been working on a better way to handle color, since ESPHome did not have color support for displays such as this; I've been working in my PR (#1050) to implement a reasonable way to describe and manage color in the config. I think I've finally got it implemented in a reasonable way, so perhaps we'll start to see more movement on these color display PRs in the coming days.

@roberttco as @Pocke01 pointed out, I believe the aim is to minimize dependencies on other libraries. Plus, we sometimes implement display features that just aren't in the Adafruit libraries, such as contrast control or display power control. That said, these libraries often provide us with a good jump start to get development of the component done more quickly.

glmnet commented:
As @kbx81 pointed out, this should be updated to handle new color apis

SenexCrenshaw commented:
I will check out @kbx81 PR.

Expaso commented:
@SenexCrenshaw, you need a hand with this? 

I happen to have one of these hooked up on my breadboard, doing..., nothing. 
Don't know if you already started refacatoring, but if not, I am willing to give it a try.

SenexCrenshaw commented:
I am lame. I will work on this today.

Expaso commented:
> I am lame. I will work on this today.

Allright! If I can do anything to help or test, let me know!

SenexCrenshaw commented:
Expaso 
Worked like a charm for me. Here is my yaml. dc_pin might be a0 on your device. 

spi:
  clk_pin: D5
  mosi_pin: D7
  miso_pin: D6

display:
- platform: st7735
  model: "INITR_18BLACKTAB"
  reset_pin: D3
  cs_pin: D8
  dc_pin: D4
  rotation: 0
  devicewidth: 128
  deviceheight: 160
  colstart: 0
  rowstart: 0
  eightbitcolor: true
  update_interval: 5s
  lambda: |-
      // Draw a line from [0,0] to [100,50]
      it.line(0, 0, 100, 50);
      // Draw the outline of a rectangle with the top left at [50,60], a width of 30 and a height of 42
      it.rectangle(50, 60, 30, 42);
      // Draw the same rectangle, but this time filled.
      it.filled_rectangle(50, 60, 30, 42);

      // Circles! Let's draw one with the center at [25,25] and a radius of 10
      it.circle(25, 25, 10);
      // ... and the same thing filled again
      it.filled_circle(25, 25, 10);

SenexCrenshaw committed:
Update to new color API and added test

SenexCrenshaw committed:
Update to new color API and added test

SenexCrenshaw committed:
Check fixes

SenexCrenshaw committed:
Check fixes

SenexCrenshaw committed:
Fixed sid length in test

SenexCrenshaw committed:
Fixed sid length in test

SenexCrenshaw committed:
Cleaned up whitespaces

SenexCrenshaw committed:
Cleaned up whitespaces

SenexCrenshaw commented:
Checks passed and its been all updated. let me know if i need to do anything else. 
@Expaso let me know if you need help getting yours to work
@kbx81 thanks for all the help!

jesserockz reviewed (dismissed):
Please fix the `tests/test1.yaml` file... You have deleted everything apart from your display...

kbx81 commented on esphome/components/st7735/st7735.cpp:
```diff
@@ -0,0 +1,474 @@
+#define ST77XX_CASET 0x2A
+#define ST77XX_RASET 0x2B
+#define ST77XX_RAMWR 0x2C
+#define ST77XX_RAMRD 0x2E
```
Avoid using #define -- these should be `const uint8_t` as you've done with most of the others. 😄

jesserockz replied:
I thought CI should have complained about these. 🤷

SenexCrenshaw replied:
i had issues with the docker checks too. it kept changing my override function names. ill clean these up.

kbx81 commented on esphome/components/st7735/st7735.h:
```diff
@@ -0,0 +1,109 @@
+#define ST77XX_CYAN 0x07FF
+#define ST77XX_MAGENTA 0xF81F
+#define ST77XX_YELLOW 0xFFE0
+#define ST77XX_ORANGE 0xFC00
```
I think we are handling colors in the `color` library. You can probably get rid of these (and the ones that follow, too). 😇

kbx81 reviewed (commented):
Left a couple comments -- nothing big! I wish I had one of these to test on! 😁 Fix up `test1.yaml`, too, and you're probably good imho. 😄

SenexCrenshaw committed:
kbx81 recommended fixes

SenexCrenshaw committed:
kbx81 recommended fixes

SenexCrenshaw committed:
test fix

SenexCrenshaw committed:
test fix

jesserockz commented:
The `static const`s are all messed up because they are missing semi colons at the end of the line

SenexCrenshaw committed:
Fixes of fixes

SenexCrenshaw committed:
Fixes of fixes

jesserockz commented:
Better =) but you still need to fix the test1.yaml file....
You have removed the tests for all of the components that were in there before...

Take a look at the [diff](https://github.com/esphome/esphome/pull/1066/files#diff-9870f3540b661068081c771f7aa9fdeb95eb1a80bd61a50e6eeeaf25105e578a)...

SenexCrenshaw committed:
Fixed test1.yaml

SenexCrenshaw committed:
Fixed test1.yaml

SenexCrenshaw committed:
Fixed test1.yaml

SenexCrenshaw committed:
Fixed test1.yaml

SenexCrenshaw commented:
@jesserockz HA. makes total sense now. it has been updated. thanks for all the help the more I understand how this flows the easier the next PR will be.

SenexCrenshaw committed:
Changed digital pin #s to gpio

SenexCrenshaw committed:
Changed digital pin #s to gpio

SenexCrenshaw committed:
Updated to match kbx's color names

SenexCrenshaw committed:
Updated to match kbx's color names

SenexCrenshaw reviewed (commented):
Made changes

SenexCrenshaw commented:
> Better =) but you still need to fix the test1.yaml file....
> You have removed the tests for all of the components that were in there before...
> 
> Take a look at the [diff](https://github.com/esphome/esphome/pull/1066/files#diff-9870f3540b661068081c771f7aa9fdeb95eb1a80bd61a50e6eeeaf25105e578a)...
> 
> I am guessing you used that file as your own test yaml file... I would suggest you put a yaml file in a `config` directory in the `esphome` folder as that is ignored by git

SenexCrenshaw reviewed (commented):
Should be good now. I didnt understand what test1.yaml was doing until now. Thanks!

Expaso commented:
@SenexCrenshaw Thanks!! Seems very close to a working version!

Found 1 issue though, I tried to use 

```
display:
- platform: st7735
  model: "INITR_MINI160x80"
```

but that doesn't pass yaml validation.
Any clue?

SenexCrenshaw committed:
Typo for ST7735_INITR_MINI_160X80

SenexCrenshaw committed:
Typo for ST7735_INITR_MINI_160X80

SenexCrenshaw commented:
> NITR_MINI160x80

typo in my code. pr was just pushed.

Expaso commented:
Hi @SenexCrenshaw , 

Thanks for the updates! I have tested them, and found a few issues around colorspace conversion.
I've created a PR for you to review: https://github.com/SenexCrenshaw/esphome/pull/11

At lease on my MINI 160x80 everything is working like a charm now!

SenexCrenshaw reviewed (commented):
Should be good to go now

SenexCrenshaw commented:
Just fyi, another PR will be coming to update the 8bit color space.

SenexCrenshaw committed:
Updated 8bit color space code
Added to_rgb_332 to color.h
fixed typo

SenexCrenshaw committed:
Updated 8bit color space code
Added to_rgb_332 to color.h
fixed typo

Expaso commented:
Wait! You need BGR332 and not RGB332, this last change will mess up the colors, although a RGB332 function could become handy in the future.

SenexCrenshaw commented:
> Wait! You need BGR332 and not RGB332, this last change will mess up the colors, although a RGB332 function could become handy in the future.

Not on my device! Is your Red and Blue swapped? I could add a setting in for that

Expaso commented:
Hmm, maybe related to: https://github.com/adafruit/Adafruit_CircuitPython_RGB_Display/pull/13

SenexCrenshaw committed:
Added in to_rgb_332,to_bgr_332, rgb_332to_rgb_556 and a more generic scale

SenexCrenshaw committed:
Added in to_rgb_332,to_bgr_332, rgb_332to_rgb_556 and a more generic scale

SenexCrenshaw committed:
Fixed MADCTL

SenexCrenshaw committed:
Fixed MADCTL

SenexCrenshaw commented:
@Expaso check out the latest PR. I scale on the fly instead of going to a lookup table. I dont expect any performance issues. I also incorporated your MADCTL change. Runs great on my device. go 8bit color and 32K of extra RAM!

SenexCrenshaw committed:
Fixp MADCTL

SenexCrenshaw committed:
Fixp MADCTL

Expaso commented:
> @Expaso check out the latest PR. I scale on the fly instead of going to a lookup table. I dont expect any performance issues. I also incorporated your MADCTL change. Runs great on my device. go 8bit color and 32K of extra RAM!

Please check my commits. The MADCTL  fix contains more then you currently have adopted. Are you testing also on on 16-bit? 

Code below still looks wrong to me because RGB and BGR should not be mixed!


  if (this->eightbitcolor_) {
    // 8-Bit color space is in **BGR332**
    const uint32_t color332 = **color.to_rgb_332();**
    uint16_t pos = (x + y * this->get_width_internal());
    this->buffer_[pos] = color332;
  } else {
    // 16-bit color-space is in **BGR565**
    const uint32_t color565 = **color.to_bgr_565();**
    uint16_t pos = (x + y * this->get_width_internal()) * 2;
    this->buffer_[pos++] = (color565 >> 8) & 0xff;
    this->buffer_[pos] = color565 & 0xff;
  }
}


I think eveything should be RGB now in the code, and fix any display issues in init-code.

SenexCrenshaw committed:
Implemented usrbgr option
updated color to support 332 bgr conversion
typo fix

SenexCrenshaw committed:
Implemented usrbgr option
updated color to support 332 bgr conversion
typo fix

SenexCrenshaw committed:
Updated to_bgr_332

SenexCrenshaw committed:
Updated to_bgr_332

SenexCrenshaw commented:
@Expaso Ok. I added in a usebgr option. My blacktab is needs rgb, i didnt get it from ada so who knows.. I figured best was to add in an option for bgr. It seems to work on my display although i dont have one that actually needs bgr. I also updated color.h to translate from BGR 332 to 565. I treated it as a 233 though... Not sure if that is correct but makes sense to me!

Please test

SenexCrenshaw committed:
Fix up for clang

SenexCrenshaw committed:
Fix up for clang

Expaso commented:
Wow thats fast! Will test!

SenexCrenshaw committed:
FIx up init code. type in buffer caused overrun

SenexCrenshaw committed:
FIx up init code. type in buffer caused overrun

SenexCrenshaw committed:
fixup protected names

SenexCrenshaw committed:
fixup protected names

SenexCrenshaw committed:
typos

SenexCrenshaw committed:
typos

SenexCrenshaw commented:
Hopefully this is it. works in 332/565 and with rgb/bgr. Note depending on your CPU you can run out of ram in you are not in 8bit (332) mode. I have to cut down rows/cols for mine to work hence 8bit!

SenexCrenshaw commented:
O and i mirrored the init from the ada fruit library.

SenexCrenshaw committed:
Matched use_bgr to its conf

SenexCrenshaw committed:
Matched use_bgr to its conf

SenexCrenshaw reviewed (commented):
Should be good to go

SenexCrenshaw committed:
color.h red fix in bgr_233to_rgb_565

SenexCrenshaw committed:
color.h red fix in bgr_233to_rgb_565

Expaso commented:
> Should be good to go

Almost, you still missed a few of my fixes in the PR:

src\main.cpp:287:38: error: 'ST7735_INITR_MINI160X80' is not a member of 'esphome::st7735'
   st7735_st7735 = new st7735::ST7735(st7735::ST7735_INITR_MINI160X80, 80, 160, 0, 0, false, false);

SenexCrenshaw committed:
Fix ST7735_INITR_MINI_160X80

SenexCrenshaw committed:
Fix ST7735_INITR_MINI_160X80

SenexCrenshaw commented:
> ST7735_INITR_MINI160X80
How does that keep sneaking back in.

Expaso commented:
> > ST7735_INITR_MINI160X80
> > How does that keep sneaking back in.

Something wrong with the merges? I fixed it in the PR a while ago.

Expaso commented:
16-bit is fine now.

8-bit is not. 1 bit of RGB332 is leaking into the wrong color channel.

Expaso commented:
Is this correct?  Comments does not seem to be in line wth code:

```
static uint32_t bgr_233to_rgb_565(uint8_t bgr233) {
    uint16_t red, green, blue;

    blue = (bgr233 & 0xc0) >> 6;  // bgr233 2 green bits now right justified
    blue = esp_scale(blue, 3, 31);
    blue = blue << 11;  // blue bits now 5 MSB bits

    green = (bgr233 & 0x38) >> 2;  // bgr233 3 green bits now right justified
    green = esp_scale(green, 7, 63);
    green = green << 5;  // green bits now 6 "middle" bits

    red = bgr233 & 0x07;  // rgb332 2 blue bits are right justified
    red = esp_scale(red, 7, 31);

    return (uint16_t)(blue | green | red);
  }
```

Expaso commented:
On a sidenote, I don't think the whole BGR/RGB conversion you added is needed at all. 
You could simply change the whole MADCTL init routing to this:

```
  uint8_t data = 0;
  if (this->model_ != INITR_HALLOWING) {
      uint8_t data = data = ST77XX_MADCTL_MX | ST77XX_MADCTL_MY;
  }
  if (this->usebgr_) {
    data = data | ST7735_MADCTL_BGR;
  } else {
    data = data | ST77XX_MADCTL_RGB;
  }
  sendcommand_(ST77XX_MADCTL, &data, 1);
```
And always use RGB in te rest of the code.

SenexCrenshaw commented:
@Expaso 
You are right about the leak. I though if the device was init to use BGR it expected BGR sent to it.. I can update the init code to what you have (cleaner) and thank for the bgr fix. I will leave those in color.h unless someone doesnt feel we dont need those functions. Im still surprised people dont use 8bit more for these devices esp with a 8266.

SenexCrenshaw committed:
Renamed bgr_233to_bgr_565 to match its function
Color space leak in bgr_233to_bgr_565.
cleaned up init code for displays.

SenexCrenshaw committed:
Renamed bgr_233to_bgr_565 to match its function
Color space leak in bgr_233to_bgr_565.
cleaned up init code for displays.

SenexCrenshaw committed:
Fix

SenexCrenshaw committed:
Fix

SenexCrenshaw committed:
clang fix

SenexCrenshaw committed:
clang fix

Expaso commented:
@SenexCrenshaw  Yes, this is it! I'm very happy with the result!

Expaso commented:
> You are right about the leak. I though if the device was init to use BGR it expected BGR sent to it.. I can update the init code to what you have (cleaner) and thank for the bgr fix. I will leave those in color.h unless someone doesnt feel we dont need those functions. Im still surprised people dont use 8bit more for these devices esp with a 8266.

Agreed, those rgb/bgr functions could become quite handy there in color.h, so I also would plead to leave them.

About the (awesome working) 8-bit mode: I even have a branch here that uses no buffer at all, since the screen already has it's own  framebuffer. But that implies you also have to wipe (parts of) the screen if you want to update it yourself.
The draw-pixel function directly calls the SPI methods in tat case.

About this PR: I think you can summon the gods to pull this in?

kbx81 commented on esphome/core/color.h:
```diff
@@ -141,6 +142,48 @@ struct Color {
+    red = esp_scale(red, 7, 31);
+
+    return (uint16_t)(blue | green | red);
+  }
```
I have some reservations about the structure of these conversion functions -- if only as it relates to consistency with the rest of what's in this library.

The `to_xxx_yyy()` functions return a value computed based on class members/variables -- `red`, `green`, `blue`, and/or `white`. These functions...don't. At very least they should be flagged `const` since they don't actually modify any of the data in the class variables, but I think it would be more consistent with the design pattern to have functions that convert the data in the class variables from one format to the other. That said, we'd need a way to initialize a `Color` object from these other formats (565/332/233) to do so. Maybe something like:

```
Color my332(0xe5);
auto color565 = my332.from_332_to_565();
```

...if that makes any sense at all. I realize that something like `uint32_t bgr_233to_bgr_565(uint8_t bgr233)` is a bit easier to write, but it kind of defeats the purpose of the class and could arguably just as well live outside of the class.

I'm not terribly attached to any of this, I'm just thinking about it in terms of consistency with what's already here as well as common design patterns in general. I'd love to hear the thoughts of others here!

P.S.: I kept typing `class` but meant `struct`. Sorry!

P.P.S.: Barring everything else I typed above, they should at least be named `rgb_332_to_rgb_565`/`bgr_233_to_bgr_565` instead of `rgb_332to_rgb_565`/`bgr_233to_bgr_565`. 😁

Expaso replied:
From a design perspective you are totally right, and constructors accepting a RGB332 or RGB565 would also be nice.

But.. from a performance perspective, if you need to convert from RGB332 to RGB565 (what we are doing here) this would bring in twice the math, because the values needs to be scaled twice (from RGB332 to RGB888 to RGB565).
 
This function is called for every pixel on the display, multiplied by every frame. 

From that perspective I would opt for putting these methods outside of the class, but keep single conversion.

SenexCrenshaw replied:
Yeah I agree with both. I am up for a separate helper class but I wouldn't think the extra overhead of using the color class as described would be impactful. 

@kbx81 you make the call. You started this whole color thing anyway ;)

jesserockz replied:
My 2c: I tend to lean towards agreeing with @kbx81 on adding the functions to the color class/struct and using them like that.

kbx81 replied:
If you are curious/concerned about the performance, write the additional function(s) and copy/paste it/them into [Compiler Explorer](https://godbolt.org) and compare the compiled assembly code. Then you can see the comparison in terms of instructions/cycles consumed. I bet they're pretty similar. The optimizer is real. 😄

Expaso replied:
Or, I will measure the real performance on the ESP8266 using my scope. Im against premature optimization though, so if this performs well enough, I would favor architechture over extreme prrformance. But we are running on microcontrollers...

kbx81 replied:
I think that's what I'm getting at -- I'm not expecting a huge difference in performance either way. Bit rotations (which is all we're really doing here) are single-cycle and they can be combined with loads/stores. If we do this and performance is notably worse, I'd be interested to look at the compiled assembly and compare to see why. Everything I know tells me it should be more or less a wash.

See also: https://developer.arm.com/documentation/ddi0439/b/Programmers-Model/Instruction-set-summary/Cortex-M4-instructions?lang=en

kbx81 replied:
Neat to see: https://godbolt.org/z/nKxM48

Expaso replied:
Well, let's see what the measurements say:

This is the transfer of 1 frame to the display, in 16-bit mode, without color conversion:
![image](https://user-images.githubusercontent.com/26247804/97548987-233ca100-19d0-11eb-97e5-93dfadf71a16.png)

This is the transfer of 1 frame to the display, in 8-bit color, using a lookup table for conversion:
![image](https://user-images.githubusercontent.com/26247804/97549499-bf66a800-19d0-11eb-8f04-7a62d53e5ce0.png)

This is the transfer of 1 frame to the display, using the currently implemented RGB332_to_565 direct conversion:
![image](https://user-images.githubusercontent.com/26247804/97549734-0e144200-19d1-11eb-8189-b803603b566f.png)

This is the transfer of 1 frame to the display, using the proposed double-conmversion (thus constructing a Color struct from RGB332, and then convert it to RGB565 again:
![image](https://user-images.githubusercontent.com/26247804/97559965-6271ee80-19de-11eb-8a1d-794277467111.png)

Conclusion: 

So indeed the double conversion adds 4.5ms overhead per frame, which is perfectly acceptable.
As you can see, using 8-bit color is fairly heavy compared to native 16-bit, but you'll gain RAM. 

@kbx81 Thanks for the info! i also looked at the ASM generated, and it was as expected. 
I'm with you. For architectural-sake we should use a `.Color(uint8_t rgb332)` constructor, and simply call `.to_rgb565()`

So thanks for the discussion guys! Learned a lot!

SenexCrenshaw replied:
I can add in the changes to this PR. 
Does anyone want to share their test code? to make my changes easier?

I know this whole 8bit thing can seem like its not needed but on a 8266 i couldnt even run this display in full screen with esphome just doing basic things. its only 80k!

Ill be away this weekend and will update this earlier next week. Great conversation!!

kbx81 replied:
I have a couple ideas to make `Color(uint32_t colorcode)` more useful. I'll see if I can cobble something together this weekend.

kbx81 replied:
Follow up: thank you @SenexCrenshaw for reading my mind and implementing everything I was dreaming about. 😉😂

kbx81 reviewed (commented):
Just leaving some thoughts on `Color` -- would love to hear from others.

SenexCrenshaw committed:
Started Color Conversion

SenexCrenshaw committed:
Started Color Conversion

SenexCrenshaw committed:
Added various bit color functions
add triadto

SenexCrenshaw committed:
Added various bit color functions
add triadto

SenexCrenshaw committed:
lint changes

SenexCrenshaw committed:
lint changes

SenexCrenshaw committed:
Various fixes

SenexCrenshaw committed:
Various fixes

SenexCrenshaw commented:
@kbx81 and I got together and came up with this. 
New color constructor for inputting colors according
EX:
Color(colorcode332, 3, 3, 2, true);
Then 2 helpers functions
triad_to16()
triad_to8()
so to output a RGB565 color. 
auto color565=color888.triad_to16(5, 6, 5);

This way you can define any scenario you want. the bool defines how the colorcode enter is "packed".
true is right aligned as such: 00000000 00000000 RRRRRGGG GGGBBBBB
false is this: 00000000 000RRRRR 00GGGGGG 000BBBBB

My display driver uses these new functions:
auto color = Color(this->buffer_[index + line], 3, 3, 2, true).triad_to16(5, 6, 5);
Changed the RGB332 color in the buffer to RGB556.

SenexCrenshaw committed:
Various formatting fixes. Wish my checks worked!

SenexCrenshaw committed:
Various formatting fixes. Wish my checks worked!

kbx81 commented:
Love it 😁

Expaso commented:
Oeh nice!

If we go this route, may I suggest an extension then?
The color struct already contains a fourth byte, currently used for the white channel. We could expose this channel, so you could use it as alpha channel. That way, image manipulation becomes possible. 
Basically, we support all these formats then: https://www.kernel.org/doc/html/v4.9/media/uapi/v4l/pixfmt-packed-rgb.html

BTW, how is BGR conversion arranged now? The new methods all seem RGB only.

kbx81 commented:
I had thought about support for alpha as I was originally working on this library, but opted against it if only for sake of simplicity. It's definitely possible, but arguably starting to get a bit outside of the scope of this PR. Maybe we spin up a new PR for that? I think it's a reasonable idea but has the potential to get...well, outside of the implementation of a display device. 😄
(Semi-related: we don't do image processing on the ESP32 -- `pillow` does that for us _before_ the image is stored into memory. When you bring alpha into the picture (I swear there's a pun here somewhere...) you're starting to get into image processing/manipulation and, well, I do worry a bit about CPU and memory constraints here...but, again, this is outside the scope of this PR.)

> BTW, how is BGR conversion arranged now? The new methods all seem RGB only.
As implemented here, it becomes irrelevant; you specify what bits you want where when you call the function(s). This is why we used `Triad` in the name -- it's less specific in that it just refers to a group of three pixels, not their arrangement (in memory or otherwise).

kbx81 commented:
Actually, taking another look, you're right -- the conversion functions lack a way to specify RGB/BRG/GRB ordering. We should fix that...

Expaso commented:
Enum would do the job just fine. 
In my opinion, even an enum with then most commonly used RGBxxx and BGRxxx formats should suffice.

SenexCrenshaw commented:
yeah i agree. ill put something together.

SenexCrenshaw committed:
Updated color api to support different formats
removed to_rgb_565

SenexCrenshaw committed:
Updated color api to support different formats
removed to_rgb_565

SenexCrenshaw committed:
lint clang fixes

SenexCrenshaw committed:
lint clang fixes

SenexCrenshaw committed:
Test1 fix

SenexCrenshaw committed:
Test1 fix

SenexCrenshaw committed:
test1.yaml fix

SenexCrenshaw committed:
test1.yaml fix

SenexCrenshaw committed:
fixed 565 in ILI9341Display

SenexCrenshaw committed:
fixed 565 in ILI9341Display

SenexCrenshaw commented:
Alright. this should cover us. RGB/BGR/GRB 888/565/332

I didnt do anything with incorporating an alpha channel. This is a display driver gone a little sideways.... 

Hopefully this is OK to merge now and thanks to everyone for all the input. I learned a lot!

SenexCrenshaw committed:
Added CodeOwners

SenexCrenshaw committed:
Added CodeOwners

jesserockz commented on esphome/core/color.h:
```diff
@@ -141,16 +199,87 @@ struct Color {
+        break;
+    }
+
+    red_color = (esp_scale8(this->red, ((1 << first_bits) - 1)) << (16 - first_bits));
+    green_color = (esp_scale8(this->green, ((1 << second_bits) - 1)) << (16 - first_bits - second_bits));
+    blue_color = esp_scale8(this->blue, (1 << third_bits) - 1);
+
+    switch (color_order) {
+      case COLOR_ORDER_RGB:
+        return red_color | green_color | blue_color;
+      case COLOR_ORDER_BGR:
+        return blue_color | green_color | red_color;
+      case COLOR_ORDER_GRB:
+        return green_color | red_color | blue_color;
+    }
```
Just trying to understand this code (and same above in `triad_to8`) 
If I am reading this correctly, the `color_order` makes no difference as the colors have been bit shifted into non-conflicting spots already and will always be RGB?
Seems like for this to work, the colors need to be bit shifted inside each case so they are in the right position?


Also it makes no sense to have `COLOR_BITNESS_888` and `COLOR_BITNESS_332` in the `triad_to16` function?
Same for `COLOR_BITNESS_888` and `COLOR_BITNESS_332` in the `triad_to8` function?

SenexCrenshaw replied:
You are right. The previous PR I had you could accomplish RGB/BGR/GBR and all the bitness on your own. I added the ENUMsto make things more readable. I totally got carried away and you are so right about those triads. 888 was added because I was considering doing something with defaults but isnt needed. I think my late night thought (or was it early morning) was trying to create flexible functions but missed the obvious. I propose changing those triads to triad_to332 and triad_to556 or just to to332/to556. 

For the color_order in the triads... If my device needs BGR I will set that in the `draw_absolute_pixel_internal` function. I receive a color and then:
`const uint32_t color332 = color.triad_to8(Color::COLOR_ORDER_RGB, Color::COLOR_BITNESS_332);`
When `update` is called 
`auto color = Color(this->buffer_[index + line], Color::COLOR_ORDER_RGB, Color::COLOR_BITNESS_332, true)
                         .triad_to16(Color::COLOR_ORDER_BGR, Color::COLOR_BITNESS_565);`
This is why I need bitness. this also saves me 1/2 the buffer space (and the whole reason I got into 332 in the first place). My 8266 doesnt have enough ram to do a full screen 16bit color.

I propose to change those to something like this:
`const uint32_t color332 = color.to332();`
`auto color = Color(this->buffer_[index + line], Color::COLOR_ORDER_RGB, Color::COLOR_BITNESS_332, true)                         .to565(Color::COLOR_ORDER_BGR);`

Color::COLOR_ORDER_RGB would be the default for those functions. 

Thoughts?

jesserockz reviewed (changes requested):
I see you are formatting the test yaml for some consistency, which is good, but please can you put that into a PR by itself. Thanks

SenexCrenshaw commented:
> I see you are formatting the test yaml for some consistency, which is good, but please can you put that into a PR by itself. Thanks

Will that block this merge? I dont mind doing a separate PR with the formatting added in but wasnt sure if you wanted me to do anything with the one attached to this PR.

SenexCrenshaw committed:
Updated CODEOWNERS

SenexCrenshaw committed:
Updated CODEOWNERS

SenexCrenshaw commented:
Ill fix the bad tests once we decide if my proposed changes make sense.

SenexCrenshaw committed:
changed to to332 and to565

SenexCrenshaw committed:
changed to to332 and to565

glmnet commented:
Sorry to arrive late to the party but why so much changes in color handling?

Can I get a resume of current issues with current implementation and what this solution advantages / problems are?

SenexCrenshaw commented:
> Can I get a resume of current issues with current implementation and what this solution advantages / problems are?

sure is a party. It started with just eightbit color conversion for this display driver. color.h was added and the thought was to expand that out. discussions went back and forth between performance and flexibility. I added a constructor to handle different color formats and then to332 and to565. I have a pr ready to go to fix this up. i think this got a little out of control but at the end of the day its a constructor to import different color formats to RGB and then to332() and to565().

glmnet commented:
You did not answer my question accurately.

I've seen all the history, it is still not clear to me why you need to convert from 332 to 556 but anyway, big question:

Is it possible to have this PR deal *ONLY* with the display, and have all color changes remove from this?
You can still add another lightweight to_whatever() function to the color struct in this PR in case your display needs it.

the `to_COLOR_ORDER_BITNESS()` method format allows you to handle all combinations, of course, it converts the color itself.
You could add `to_bgr_332()` and `to_rgb_332()` etc.

If for some reason you need to build a Color out of not a RGB888 color you can write some private internal helper meanwhile.
If you need to convert from 332 to 556 or whatever, and don't want to do a double conversion then you can also do that privately also.

Let's then merge this PR and get another one for doing color changes, test file organization, etc. 
In general this PR should not interfere any other display, that will require much more testing or review from other contributors.

SenexCrenshaw commented:
Originally I had it all internal and it was asked to add it to color.h 
I convert from color to 332 to store in my display buffer in case the user whats 8bit color, saving 1/2 the buffer space. a 8266 doesnt have enough ram to handle full color for a 160x120 display and all the functions i needed. So 332 goes into the buffer and 332 -> 565 to my display on write.

How about I close this PR and just re-submit a new one without all the color and test formatting changes?

glmnet commented:
> I convert from color to 332 to store in my display buffer in case the user whats 8bit color, saving 1/2 the buffer space. a 8266 doesnt have enough ram to handle full color for a 160x120 display and all the functions i needed. So 332 goes into the buffer and 332 -> 565 to my display on write.

This is a good idea, this will require you to convert from Color to 332, which in other displays it is currently done in the color.h file.
Then converting from the buffer to the values which goes to the wire it can be done on the fly I guess. I don't see a huge advantage reusing the to_556 part which will justify adding all this (this is opinion of course)

> How about I close this PR and just re-submit a new one without all the color and test formatting changes?

If that works for you, it will be ok

kbx81 commented:
@glmnet some additional background here: the motivation for adding the bit-width/arrangement conversion functions to `Color` here was so that they could live in a central place -- I see this being useful for other displays so that they wouldn't have to reinvent the wheel for each one. We already have a function in the `Color` library to convert from the (native) 888 format to the 565 format so it seemed to make sense to add functionality to convert to 332 here, as well. I was working with @SenexCrenshaw to add that support in a constructive way so that it could be reused for other displays that may later adopt this approach. That said, it _has_ dwelled a bit outside of the scope of the simple addition of support for this specific display, however I would argue that this is also the first display that adds support for this 332 format so I don't think it's _entirely_ out of line.

While not optimal quality-wise, the 332 format is pretty essential for the '8266s since they have so little memory -- they won't be able to handle larger color displays. This is likely why we had only binary display support until earlier this year when we added `Color` in #1050. I have to assume that others who want to use large color displays with an '8266 will also come to require this same conversion functionality.

Just offering my friendly two cents. 😇

glmnet commented:
I see, it's just that we are too slow reviewing stuff and adding complexity to PRs is just going to be nuts. 

I guess custom color buffers will be interesting, so user can select 2 bits per pixel, 4, 8, 16, or even crazy stuff where you have regions full color and others just two colors, what about paletted buffers? 

DisplayBuffer requires you to implement the `...absolute_draw_pixel` you can have an intermediate class which implements this, for a binary buffer, for a 4 bit buffer, color, grayscale, etc. This way you can do optimized conversion in only one place, the driver itself will need to call a function to retrieve a pixel from the buffer in the required color form e.g. 16 bits

This can be all supported via a custom buffer and advantage will be taken by all color displays, please do so in another PR though.

SenexCrenshaw committed:
Waiting for color.h changes

SenexCrenshaw committed:
Waiting for color.h changes

SenexCrenshaw committed:
Stage changes

SenexCrenshaw committed:
Stage changes

SenexCrenshaw commented:
> > I convert from color to 332 to store in my display buffer in case the user whats 8bit color, saving 1/2 the buffer space. a 8266 doesnt have enough ram to handle full color for a 160x120 display and all the functions i needed. So 332 goes into the buffer and 332 -> 565 to my display on write.
> 
> This is a good idea, this will require you to convert from Color to 332, which in other displays it is currently done in the color.h file.
> Then converting from the buffer to the values which goes to the wire it can be done on the fly I guess. I don't see a huge advantage reusing the to_556 part which will justify adding all this (this is opinion of course)
> 
> > How about I close this PR and just re-submit a new one without all the color and test formatting changes?
> 
> If that works for you, it will be ok

@glmnet I split out color.h into its own PR to support 332 color conversion and RGB/BGR/GRB formats. If this can be merged I will use it in this PR and clean things up in just my display driver.
#1381

SenexCrenshaw commented:
> I see, it's just that we are too slow reviewing stuff and adding complexity to PRs is just going to be nuts.
> 
> I guess custom color buffers will be interesting, so user can select 2 bits per pixel, 4, 8, 16, or even crazy stuff where you have regions full color and others just two colors, what about paletted buffers?
> 
> DisplayBuffer requires you to implement the `...absolute_draw_pixel` you can have an intermediate class which implements this, for a binary buffer, for a 4 bit buffer, color, grayscale, etc. This way you can do optimized conversion in only one place, the driver itself will need to call a function to retrieve a pixel from the buffer in the required color form e.g. 16 bits
> 
> This can be all supported via a custom buffer and advantage will be taken by all color displays, please do so in another PR though.

If I am reading this correctly you are suggesting a buffer class be made to handle this? That was something I considered as well but was trying not to blow up this PR anymore than it has! I will put a separate PR together for that. In the meantime I will change this PR to work with the existing color.h so we can get this one merged.
I created 2 PRs for color.h and the tests yaml formatting. For color.h I do believe these functions will be valuable for others to use even potentially outside a display driver and will use it in the buffer PR.
#1381 
#1382

SenexCrenshaw committed:
Removed all changes except this driver

SenexCrenshaw committed:
Removed all changes except this driver

SenexCrenshaw committed:
Moved color functions into driver

SenexCrenshaw committed:
Moved color functions into driver

SenexCrenshaw committed:
lint changes

SenexCrenshaw committed:
lint changes

SenexCrenshaw committed:
Lint and removed unrelated display driver changes

SenexCrenshaw committed:
Lint and removed unrelated display driver changes

SenexCrenshaw committed:
Lint changes

SenexCrenshaw committed:
Lint changes

SenexCrenshaw commented:
Finally fixed my linting issues in my dev environment. this will greatly reduce my PR errors!

SenexCrenshaw commented:
So not sure if you want to merge this if you like 
#1381 
and 
#1386 
If they are merged I will port this driver to use that. If not merge away!

SenexCrenshaw committed:
Updated with latest color api

SenexCrenshaw committed:
pulled from origin

glmnet commented:
Ok, now that the color.h got changed, In case you want to work on the buffer types, let's have this one merged first, so you can work with this display.

jesserockz commented on esphome/components/st7735/st7735.cpp:
```diff
@@ -0,0 +1,508 @@
+
+// clang-format on
+static const char *TAG = "st7735";
+inline static uint8_t esp_scale(uint8_t i, uint8_t scale, uint8_t max_value = 255) { return (max_value * i / scale); }
```
I believe this is no longer needed
```suggestion

```

SenexCrenshaw replied:
I actually had this in a PR but was waiting to see what the buffer piece was going to end up being. Sorry i didnt publish earlier!

jesserockz commented on esphome/components/st7735/st7735.cpp:
```diff
@@ -0,0 +1,508 @@
+  this->disable();
+}
+
+// uint16_t HOT ST7735::rgb_332to_rgb_556_(uint8_t rgb332) {
+//   uint16_t red, green, blue, blue2;
+
+//   red = (rgb332 & 0xe0) >> 5;
+//   red = esp_scale(red, 7, 31);
+//   red = red << 11;
+
+//   green = (rgb332 & 0x1c) >> 2;
+//   green = esp_scale(green, 7, 63);
+//   green = green << 5;
+
+//   blue = rgb332 & 0x03;
+//   blue = esp_scale(blue, 3, 31);
+
+//   return (uint16_t)(red | green | blue);
+// }
+
+// uint8_t ST7735::to_rgb_332_(Color color) {
+//   return (esp_scale8(color.red, ((1 << 3) - 1) << 5)) | (esp_scale8(color.green, ((1 << 3) - 1) << 2)) |
+//          esp_scale8(color.blue, (1 << 2) - 1);
+// }
```
No longer needed
```suggestion

```

jesserockz reviewed (changes requested):
A couple of old bits to remove I think

jesserockz commented on esphome/components/st7735/st7735.cpp:
```diff
@@ -0,0 +1,486 @@
+  this->disable();
+}
+
+
```
```suggestion

```

SenexCrenshaw commented:
@jesserockz yeah sorry. I wouldnt check this one until I hash out the buffer thing with @glmnet .

SenexCrenshaw committed:
Updated for color.h changes

SenexCrenshaw committed:
pulled test1 from dev

SenexCrenshaw commented:
should be all cleaned up now. thanks for the help

SenexCrenshaw committed:
Added test

Expaso commented:
Holy Cow what a party 🎉. 
@SenexCrenshaw I was away for a few days, taking a leak and when I returned this PR got merged! Did I miss something!? 🤔 😋
